docker/Dockerfile

```diff
@@ -0,0 +1,73 @@
+# Copyright 2015 Telefónica Investigación y Desarrollo, S.A.U
+#
+# This file is part of the IoT Agent for the Ultralight 2.0 protocol (IOTAUL) component
+#
+# IOTAUL is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the License,
+# or (at your option) any later version.
+#
+# IOTAUL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with IOTAUL.
+# If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by the GNU Affero General Public License
+# please contact with: [daniel.moranjimenez@telefonica.com]
+
+ARG NODE_VERSION=8.12.0-slim
+FROM node:${NODE_VERSION}
+ARG GITHUB_ACCOUNT=telefonicaid
+ARG GITHUB_REPOSITORY=iotagent-ul
+ARG DOWNLOAD_TYPE=latest
+
+ENV GITHUB_ACCOUNT=${GITHUB_ACCOUNT}
+ENV GITHUB_REPOSITORY=${GITHUB_REPOSITORY}
+ENV DOWNLOAD_TYPE=${DOWNLOAD_TYPE}
+
+MAINTAINER FIWARE IoTAgent Team. Telefónica I+D
+
+#
+# The following line retrieves the latest source code from GitHub.
+# 
+# To obtain the latest stable release run this Docker file with the parameters
+# --no-cache --build-arg DOWNLOAD_TYPE=stable
+#
+# Alternatively for local development, just copy this Dockerfile into file the
+# root of the repository and copy over your local source using : 
+#
+# COPY . /opt/iotaul/
+#
+RUN if [ ${DOWNLOAD_TYPE} = "latest" ] ; then RELEASE="master"; else RELEASE=$(curl -s https://api.github.com/repos/${GITHUB_ACCOUNT}/${GITHUB_REPOSITORY}/releases/latest | grep 'tag_name' | cut -d\" -f4); fi && \
```

Comment:
Could this RUN step be merge with the othe one? I undertand doing so would have two advantages:

1. Only one layer in docker image building
2. To remove some redundant steps such as `apt-get clean`, etc.

Reply:
The second `RUN` statement occurs after the `WORKDIR` has been set, hence it is running from a separate directory. The download is effectively just copying a zip to root, exploding it and moving it to the right place. It is the logical equivalent to the `COPY` statement so I've kept it as a separate layer.

As you know I'm trying to do three incompatible things here, and I think that keeping a separate layer equivalent to the `COPY` statement is the simplest way to  be able to maintain a single `Dockerfile` for Automated Builds which can be painlessly altered for local use.

Combining the two would mean amending the locations the various bash command are running in, and I'd see that as an optimization too far at this stage. It would also complicate the creation of a local use _"Hacker's"_ `Dockerfile`

*If* there are two layers, the clean-up statements need to be repeated as you can only clean up within the current layer.

Reply:
4% is fine. Thanks for the info.

NTC